source/symbols/TypeHelpers.h

```diff
@@ -0,0 +1,208 @@
+        default:
+            return nullptr;
+    }
+}
+
+using PackVector = decltype(std::declval<ConstantValue>().bitstream());
+static SVInt slicePacked(PackVector::const_iterator& iter, bitwidth_t& bit, bitwidth_t width) {
+    const ConstantValue* cp;
+    for (;;) {
+        cp = *iter;
+        if (cp->isInteger()) {
+            if (bit < cp->integer().getBitWidth())
+                break;
+        }
+        else {
+            if (bit < cp->str().length() * CHAR_BIT)
+                break;
+        }
+        bit = 0;
+        iter++;
+    }
+    if (cp->isInteger()) {
+        const auto& ci = cp->integer();
+        auto msb = ci.getBitWidth() - bit - 1;
+        auto lsb = std::min(bit + width, ci.getBitWidth());
+        lsb = ci.getBitWidth() - lsb;
+        bit += msb - lsb + 1;
+        ASSERT(bit <= ci.getBitWidth());
+        return ci.slice(static_cast<int32_t>(msb), static_cast<int32_t>(lsb));
+    }
+    else {
+        std::string_view str = cp->str();
+        auto byte0 = bit / CHAR_BIT;
+        auto byte1 = (bit + width - 1) / CHAR_BIT;
+        bitwidth_t len;
+        if (byte1 < str.length())
+            len = byte1 - byte0 + 1;
+        else {
+            len = static_cast<bitwidth_t>(str.length() - byte0);
+            width = len * CHAR_BIT - bit % CHAR_BIT;
+        }
+        SmallVectorSized<byte, 8> buffer;
+        const auto substr = str.substr(byte0, len);
+        for (auto it = substr.rbegin(); it != substr.rend(); it++)
+            buffer.append(static_cast<byte>(*it));
+        len *= CHAR_BIT;
+        auto ci = SVInt(len, span<const byte>(buffer), false);
+        auto msb = len - bit % CHAR_BIT - 1;
+        auto lsb = CHAR_BIT - 1 - (bit + width - 1) % CHAR_BIT;
+        bit += width;
+        ASSERT(bit <= str.length() * CHAR_BIT);
+        if (lsb > 0 || msb < len - 1)
+            return ci.slice(static_cast<int32_t>(msb), static_cast<int32_t>(lsb));
+        else
+            return ci;
+    }
+}
+
+/// Performs unpack operation on a bit stream
+static ConstantValue unpack(const Type& type, PackVector::const_iterator& iter, bitwidth_t& bit,
+                            bitwidth_t& dynamic) {
+    if (type.isIntegral()) {
+        auto width = type.getBitWidth();
+        SmallVectorSized<SVInt, 8> buffer;
+        while (width > 0) {
+            auto ci = slicePacked(iter, bit, width);
+            ASSERT(ci.getBitWidth() <= width);
+            width -= ci.getBitWidth();
+            if (!type.isFourState())
+                ci.flattenUnknowns();
+            buffer.emplace(ci);
+        }
+        auto cc = SVInt::concat(span<SVInt const>(buffer.begin(), buffer.end()));
+        cc.setSigned(type.isSigned());
+        return cc;
+    }
+    if (type.isString()) {
+        if (!dynamic)
+            return std::string();
+        auto width = dynamic;
+        ASSERT(width % CHAR_BIT == 0);
+        dynamic = 0;
+        SmallVectorSized<SVInt, 8> buffer;
```

Comment:
This code is fully duplicated several times; please factor out into a helper method (or local lambda).

Reply:
Yes, I will do that.